File: 조현영/3주차/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <climits> // 자료형의 최댓값을 이용함

using namespace std;

long long min(long long a, long long b){
    return a >=b ? b : a;
}

long long solution(int n, vector<int> times) {
    long long answer = LLONG_MAX; //long long형의 최댓값
    sort(times.begin(), times.end()); // 최댓값을 구하기 위해 자료형 정렬
    long long start = 1;
    long long end = (long long)n * (long long)times[times.size()-1];

    while(start <= end){
        long long mid = (start+end)/2;
        long long temp = 0;

        for(int i=0; i<times.size(); i++){
            temp += mid/times[i];
        }

        if(temp >= n){
            answer = min(answer, mid);
            end = mid-1;
        }
        else{
            start = mid+1;
        }
    }
    return answer;
}

File: 박가영/3주차/입국심사.py
def solution(n, times): # 이진탐색
    answer = 0
    
    times.sort()
    left = times[0]
    right = times[len(times)-1] * n #최대 소요 시간

    while left <= right:
        mid = (start + end)//2
        total = 0 # 심사관이 mid분 동안 심사한 사람수

        for time in times:
            total += mid // time
        
        if total >= n:
            answer = mid
            right = mid - 1
        elif total < n:
            left = mid + 1

    return answer